lane.py

```diff
@@ -2138,99 +1664,325 @@ def _featured_works_with_lanes(
+        audience doesn't match what we're looking for.
+        """
+        if not self.audiences:
+            return []
+        return [Work.audience.in_(self.audiences)]
 
-        :param unused_by_tier: A dictionary mapping quality tiers to
-        lists of unused MaterializedWorkWithGenre items. Because the
-        same book may have shown up as multiple
-        MaterializedWorkWithGenre items, it may show up as 'unused'
-        here even if another occurance of it has been used.
+    def customlist_filter_clauses(self, qu):
+        """Create a filter clause that only books that are on one of the
+        CustomLists allowed by Lane configuration.
 
-        :param used_by_tier: A dictionary mapping quality tiers to lists
-        of previously used MaterializedWorkWithGenre items. These will only
-        be chosen once every item in unused_by_tier has been chosen.
+        :return: A 3-tuple (query, clauses).
 
-        :param previously_used: A set of work IDs corresponding to
-        previously selected MaterializedWorkWithGenre items. A work in
-        `unused_by_tier` will be treated as actually having been used
-        if its ID is in this set.
+        `query` is the same query as `qu`, possibly extended with
+        additional table joins.
 
+        `clauses` is a list of SQLAlchemy statements for use in a
+        filter() or case() statement.
         """
-        if not additional_needed:
-            return
-        additional_found = 0
-        for by_tier in unused_by_tier, used_by_tier:
-            # Go through each tier in decreasing quality order.
-            for tier in sorted(by_tier.keys(), key=lambda x: -x):
-                mws = by_tier[tier]
-                random.shuffle(mws)
-                for mw in mws:
-                    if (by_tier is unused_by_tier
-                        and mw.works_id in previously_used):
-                        # We initially thought this work was unused,
-                        # and put it in the 'unused' bucket, but then
-                        # the work was used after that happened.
-                        # Treat it as used and don't use it again.
-                        continue
-                    yield (mw, self)
-                    previously_used.add(mw.works_id)
-                    additional_found += 1
-                    if additional_found >= additional_needed:
-                        # We're all done.
-                        return
+        if not self.uses_customlists:
+            # This lane does not require that books be on any particular
+            # CustomList.
+            return qu, []
+
+        # We will be joining against CustomListEntry at least
+        # once. For a lane derived from the intersection of two or
+        # more custom lists, we may be joining CustomListEntry
+        # multiple times. To avoid confusion, we make a new alias for
+        # the table every time.
+        a_entry = aliased(CustomListEntry)
```

Comment:
This code (and the genre code) is simplified since we're no longer starting with the materialized view. We now have to join against `CustomListEntry` or `WorkGenre` every time, since we don't start off with a 'free' join created when the materialized view was created. This is less efficient, but again, we almost never need to use this code now.